### PR TITLE
Fix parsing of $#_

### PR DESCRIPTION
**Pull request by valcomm**
Probably would be a good idea to do more specialized global (actually general) variables parsing.  In the current state it parses some non-existing things as correct ones.
I mean splitting general variables by type and defining matching valid sigils to each group.

### CODE REVIEW

File: lib/Guacamole.pm [hunk 1]
```diff
@@ -355,6 +355,7 @@ GlobalVarExpr ::= '$#'
                 | SigilArray  GlobalVariables ElemSeq0
                 | SigilHash   GlobalVariables ElemSeq0
                 | SigilGlob   GlobalVariables ElemSeq0
+                | SigilArrayTop GlobalVariables ElemSeq0
```
Comment (vickenty) [inline lib/Guacamole.pm:358]: I don't think array top allows any element access operators to follow the identifier, so `ElemSeq0` can be removed.

File: t/Statements/Expressions/Value/NonLiteral/Variable.t [hunk 2]
```diff
@@ -91,6 +91,7 @@ parses('$^X');
 parses('@+');
 parses('@-');
 parses('@_');
+parses('$#_');
```
Comment (vickenty) [inline t/Statements/Expressions/Value/NonLiteral/Variable.t:94]: Please add tests for `$#+` and `$#-` as well.

**Review by vickenty — CHANGES_REQUESTED** [review]

**Comment by vickenty** [comment]
>  In the current state it parses some non-existing things as correct ones.
Do you have an example of an expression that perl would reject, but we parse?

File: t/Statements/Expressions/Value/NonLiteral/Variable.t (hunk 2 above)
Comment (valcomm) [reply t/Statements/Expressions/Value/NonLiteral/Variable.t:94]: Just adding tests doesn't do any good
```
t/Statements/Expressions/Value/NonLiteral/Variable.t ............... 68/? #

#   Failed test ''$#+': parsed unambiguously'
#   at t/Statements/Expressions/Value/NonLiteral/Variable.t line 92.
#          got: '0'
#     expected: '1'
#

#   Failed test ''$#-': parsed unambiguously'
#   at t/Statements/Expressions/Value/NonLiteral/Variable.t line 94.
#          got: '0'
#     expected: '1'
t/Statements/Expressions/Value/NonLiteral/Variable.t ............... 87/? # Looks like you failed 2 tests of 96.
t/Statements/Expressions/Value/NonLiteral/Variable.t ............... Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/96 subtests
```
That's surprising, as '-'/'+' are in GlobalVariables and should be parsed.

**Comment by valcomm** [comment]
> x

For example `$#0`.  What I meant is, for example, that GlobalVariables might reference real world scalar or hash, but parsing rule allows array operations on them.

File: lib/Guacamole.pm (hunk 1 above)
Comment (valcomm) [reply lib/Guacamole.pm]: Updated the commit.  Thanks for the advice

**Comment by vickenty** [comment]
> for example, that GlobalVariables might reference real world scalar or hash, but parsing rule allows array operations on them.

That's fine. As long as it doesn't affect parsing (structure that we produce), we shouldn't really care. `B::Concise` output is a good guidance here. For example, `$^M`, `@^M` and `%^M` are all accepted, even though only one will actually have value at runtime. 

On the other hand, we don't want to allow inputs that would be rejected by perl during compilation. `$#` seems to be a special case in the language, allowing only a subset of characters after it, and without any whitespace. 

It might be easier to define a new lexeme rule with all possible variants of `$#_` and reference it as a unit in `GlobalVarExpr`.

**Comment by valcomm** [comment]
Yeah, that's what I meant - a new lexeme rule.  Also for @var{} too.  Maybe some others too.  I'd split global (general) variables by type and then define what prefixes/constructs are allowed for each group.  As I see 'standard perl' is a subset, not a superset.

**Comment by jeffreykegler** [comment]
By the way, I took a look at how the bleeding edge Perl handles this, without gathering any useful insights.  The code there is anything but clean.  For example, from the comments: 

`/* '$#' is banned, while '$ # comment' isn't */`

Sigh.

**Pull request closed without merging**